Dockerfile

```diff
@@ -1,17 +1,82 @@
-FROM python:3.11-alpine
+# syntax=docker/dockerfile:1
 
-RUN apk add pipx
+#############################
+# Development Image (dev)
+#############################
+FROM python:3.12-slim AS dev
 
-WORKDIR template-python
+# Install essential packages (sudo, git, curl, build tools) with apt cache mounts.
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends sudo git curl build-essential && \
+    rm -rf /var/lib/apt/lists/*
 
-COPY . .
+# Tell Git that the workspace is safe.
+RUN git config --system --add safe.directory '*'
 
-RUN adduser -S strg \
-      && chown -R strg /template-python
+# Create a non-root user with passwordless sudo (for VS Code Dev Container compatibility).
+RUN groupadd --gid 1000 user && \
+    useradd --create-home --no-log-init --uid 1000 --gid 1000 --shell /bin/bash user && \
+    echo 'user ALL=(ALL) NOPASSWD:ALL' | tee /etc/sudoers.d/user && \
+    chmod 0440 /etc/sudoers.d/user
 
-USER strg
+USER user
 
-ENV PATH="/home/strg/.local/bin:$PATH"
-RUN pipx ensurepath && pipx install poetry==1.8.3 && poetry install
+# Configure the non-root user's shell (history and Starship prompt).
+RUN mkdir -p /home/user/.history && \
+    echo 'HISTFILE=/home/user/.history/.bash_history' >> /home/user/.bashrc && \
+    echo 'bind "\"\e[A\": history-search-backward"' >> /home/user/.bashrc && \
+    echo 'bind "\"\e[B\": history-search-forward"' >> /home/user/.bashrc && \
+    echo 'eval "$(starship init bash)"' >> /home/user/.bashrc
 
-RUN poetry run python -m unittest discover -s "test" -p "*test*.py"
+# Set the working directory as expected by VS Code Dev Containers.
+WORKDIR /workspaces/python-app-template
+
+# Install Poetry using pip.
+RUN curl -sSL https://install.python-poetry.org | python3 -
+ENV PATH="/home/user/.local/bin:$PATH" \
+    POETRY_VIRTUALENVS_IN_PROJECT=1 \
+    POETRY_NO_INTERACTION=1
+
+# Copy dependency specification files first to leverage Docker layer caching.
+COPY --chown=user:user pyproject.toml poetry.lock* ./
+
+# Install all dependencies (including dev dependencies) using Poetry.
+# A cache mount is used to speed up repeated installs.
+RUN --mount=type=cache,target=/home/user/.cache/pypoetry \
+    poetry install
+
+# Copy the remainder of the source code.
+COPY --chown=user:user . .
+
+# Ensure the virtual environment's binaries are in PATH.
+ENV PATH="/workspaces/python-app-template/.venv/bin:$PATH"
+
+# For VS Code dev containers, the default command is to launch an interactive shell.
+CMD ["bash"]
+
+#############################
+# Runtime Image (app)
```

Comment:
Maybe add a commented `EXPOSE <port>` instruction so that it's not forgotten to specify on which port an app is listening (if any).